sim/firesim-lib/src/main/scala/bridges/DromajoBridge.scala

```diff
@@ -0,0 +1,178 @@
+
+class DromajoBridgeModule(key: DromajoKey)(implicit p: Parameters) extends BridgeModule[HostPortIO[DromajoTargetIO]]()(p)
+{
+  lazy val module = new BridgeModuleImp(this) with UnidirectionalDMAToHostCPU {
+    // CONSTANTS: DMA Parameters
+    lazy val QUEUE_DEPTH = 64
+    lazy val TOKEN_WIDTH = 512 // in b
+    lazy val toHostCPUQueueDepth = QUEUE_DEPTH
+    lazy val dmaSize = BigInt((TOKEN_WIDTH / 8) * QUEUE_DEPTH)
+    // CONSTANTS
+
+    // setup io
+    val io = IO(new WidgetIO)
+    val hPort = IO(HostPort(new DromajoTargetIO(key.insnWidths, key.vecSizes)))
+
+    // the target is ready to both send/receive data
+    val tFire = hPort.toHost.hValid && hPort.fromHost.hReady
+
+    // helper to get number to round up to nearest multiple
+    def roundUp(num: Int, mult: Int): Int = { (((num - 1) / mult) + 1) * mult }
+
+    // get the traces
+    val traces = hPort.hBits.trace.insns.map({ unmasked =>
+      val masked = WireDefault(unmasked)
+      masked.valid := unmasked.valid && !hPort.hBits.trace.reset
+      masked
+    })
+    private val iaddrWidth = roundUp(traces.map(_.iaddr.getWidth).max, 8)
+    private val insnWidth  = roundUp(traces.map(_.insn.getWidth).max, 8)
+    private val wdataWidth = roundUp(traces.map(_.wdata.getWidth).max, 8)
+    private val causeWidth = roundUp(traces.map(_.cause.getWidth).max, 8)
+    private val tvalWidth  = roundUp(traces.map(_.tval.getWidth).max, 8)
+
+    // hack since for some reason padding a bool doesn't work...
+    def boolPad(in: Bool, size: Int): UInt = {
+      val temp = Wire(UInt(size.W))
+      temp := in.asUInt
+      temp
+    }
+
+    val paddedTraces = traces.map { trace =>
+      Cat(trace.tval.pad(tvalWidth),
+        Cat(trace.cause.pad(causeWidth),
+          Cat(boolPad(trace.interrupt, 8),
+            Cat(boolPad(trace.exception, 8),
+              Cat(trace.priv.asUInt.pad(8),
+                Cat(trace.wdata.pad(wdataWidth),
+                  Cat(trace.insn.pad(insnWidth),
+                    Cat(trace.iaddr.pad(iaddrWidth), boolPad(trace.valid, 8)))))))))
+    }
+
+    val maxTraceSize = paddedTraces.map(t => t.getWidth).max
+    val outDataSzBits = outgoingPCISdat.io.enq.bits.getWidth
+
+    // constant
+    val totalStreamsPerToken = 2 // minTraceSz==190b so round up to nearest is 256b
+                                 // constant
+
+    require(maxTraceSize < (outDataSzBits / totalStreamsPerToken), "All instruction trace bits must fit in 256b")
+
+    // how many streams being sent over
+    val numStreams = traces.size
+    // num tokens needed to display full stream
+    val numTokenForAll = ((numStreams - 1) / totalStreamsPerToken) + 1
+
+    // tracequeue is full
+    val traceQueueFull = outgoingPCISdat.io.deq.valid && !outgoingPCISdat.io.enq.ready
+
+    // only inc the counter when the something is sent... input is valid... and output is avail
+    val counterFire = outgoingPCISdat.io.enq.fire() && hPort.toHost.hValid && !traceQueueFull
```

Comment:
Aren't the latter two terms redundant there?
outgoingPCISdat.io.enq.fire() --> io.enq.valid --> toHost.hValid
outgoingPCISdat.io.enq.fire() --> io.enq.ready --> !traceQueueFull

Reply:
Yes. I think there is some redundancy. However, I think this makes it explicitly clear that the counter must fire with these three conditions (even if they reduce down to a simpler set of signals). I viewed this Bridge as a "more in-depth making Bridge tutorial" so I maybe I am being a bit too verbose.

Reply:
If you want to be explicit, would be better instead to write the simplest expression for the fire logic, and then add chisel assertions for the implications. 
i.e., assert(!counterFire || toHost.hValid, "Payload enqueued without valid trace token")  

As someone else reading your code, when i see the more complex expression i have to make sure i understand what each term is contributing, especially since these handshaking expressions tend to be the most common source of simulation non-determinism. 

I'll let you make the call on this one.

Reply:
I just simplified it and added a bit more to the comment.